Framework/DataHandling/src/LoadEventNexus.cpp

```diff
@@ -625,6 +637,115 @@ std::shared_ptr<BankPulseTimes> LoadEventNexus::runLoadNexusLogs(
   return out;
 }
 
+/** Load the instrument from the nexus file
+ *
+ * @param nexusfilename :: The name of the nexus file being loaded
+ * @param localWorkspace :: Templated workspace in which to put the instrument
+ *geometry
+ * @param alg :: Handle of the algorithm
+ * @param returnpulsetimes :: flag to return shared pointer for
+ *BankPulseTimes, otherwise NULL.
+ * @param nPeriods : Number of periods (write to)
+ * @param periodLog : Period logs DateAndTime to int map.
+ * @param allow_list: list of properties that will be loaded
+ * @param block_list: list of properties that will be excluded from loading
+ *
+ * @return Pulse times given in the DAS logs
+ */
+template <typename T>
```

Comment:
This method has 85 statements of which 83 are a reproduction of  `runLoadNexusLogs(nexusfilename, localWorkspace, alg, returnpulsetimes, nPeriods, periodLog)`. Why not just add two new default arguments to the original `runLoadNexusLogs` method?
```cpp
  template <typename T>
  static std::shared_ptr<BankPulseTimes> runLoadNexusLogs(
      const std::string &nexusfilename, T localWorkspace, Algorithm &alg,
      bool returnpulsetimes, int &nPeriods,
      std::unique_ptr<const Kernel::TimeSeriesProperty<int>> &periodLog,
      const std::vector<std::string> &allow_list = {},
      const std::vector<std::string> &block_list = {});
```

Reply:
Because there are other files somehow tied to the current signature, and I don't want this change to have way too many side effects.
So to maintain backward compatibility (as in not breaking other people stuff that are depending on this function), I overloaded it instead.

Reply:
Just to be 100% clear on this: did you introduce these two arguments as _default arguments_ ? (e.g. `const std::vector<std::string> &allow_list={}`)
If the arguments have default value, I don't see why [this line](https://github.com/mantidproject/mantid/blob/b70d0b9224eaf9ab542c80556fbe08b91cc0d118/Framework/DataHandling/src/LoadTOFRawNexus.cpp#L510https://github.com/mantidproject/mantid/blob/b70d0b9224eaf9ab542c80556fbe08b91cc0d118/Framework/DataHandling/src/LoadTOFRawNexus.cpp#L510) would not compile :frowning:

Reply:
TL;DR;
=====
We are dealing with a template function, not a regular function, therefore the way argument list being handled is slightly different from what you would expected from Python.

Longer version
===========
I did a quick test by adding a new dummy arguments `const std::vector<std::string> &test = {}` to the new function following you suggestion, and here is the error message
```
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:656:33: error: redeclaration of ‘static std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&)’ may not have default arguments [-fpermissive]
  656 | std::shared_ptr<BankPulseTimes> LoadEventNexus::runLoadNexusLogs(
      |                                 ^~~~~~~~~~~~~~
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:851:46: error: default argument specified in explicit specialization [-fpermissive]
  851 |     const std::vector<std::string> &test = {}) {
      |                                              ^
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp: In static member function ‘static std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&) [with T = std::shared_ptr<Mantid::DataHandling::EventWorkspaceCollection>; std::string = std::__cxx11::basic_string<char>]’:
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:851:37: warning: unused parameter ‘test’ [-Wunused-parameter]
  851 |     const std::vector<std::string> &test = {}) {
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp: In instantiation of ‘static std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&) [with T = std::shared_ptr<Mantid::API::MatrixWorkspace>; std::string = std::__cxx11::basic_string<char>]’:
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:855:17:   required from here
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:662:37: warning: unused parameter ‘test’ [-Wunused-parameter]
  662 |     const std::vector<std::string> &test = {}) {
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
```
> Notice the warning `unused parameter ‘test’ [-Wunused-parameter]` at the end.  It will trigger a build server error even if we magically get the default argument to work with this template function.

Now if I removed the default argument in the cpp but leave the declaration in the template def in the header, here is the error message
```
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:656:33: error: no declaration matches ‘std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&)’
  656 | std::shared_ptr<BankPulseTimes> LoadEventNexus::runLoadNexusLogs(
      |                                 ^~~~~~~~~~~~~~
In file included from /home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:7:
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/inc/MantidDataHandling/LoadEventNexus.h:104:42: note: candidates are: ‘template<class T> static std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&)’
  104 |   static std::shared_ptr<BankPulseTimes> runLoadNexusLogs(
      |                                          ^~~~~~~~~~~~~~~~
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/inc/MantidDataHandling/LoadEventNexus.h:98:42: note:                 ‘template<class T> static std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&)’
   98 |   static std::shared_ptr<BankPulseTimes> runLoadNexusLogs(
      |                                          ^~~~~~~~~~~~~~~~
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/inc/MantidDataHandling/LoadEventNexus.h:72:17: note: ‘class Mantid::DataHandling::LoadEventNexus’ defined here
   72 | class DLLExport LoadEventNexus : public API::NexusFileLoader {
      |                 ^~~~~~~~~~~~~~
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:843:1: error: template-id ‘runLoadNexusLogs<Mantid::DataHandling::EventWorkspaceCollection_sptr>’ for ‘std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, Mantid::DataHandling::EventWorkspaceCollection_sptr, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&)’ does not match any template declaration
  843 | LoadEventNexus::runLoadNexusLogs<EventWorkspaceCollection_sptr>(
      | ^~~~~~~~~~~~~~
In file included from /home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:7:
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/inc/MantidDataHandling/LoadEventNexus.h:104:42: note: candidates are: ‘template<class T> static std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&, const std::vector<std::__cxx11::basic_string<char> >&)’
  104 |   static std::shared_ptr<BankPulseTimes> runLoadNexusLogs(
      |                                          ^~~~~~~~~~~~~~~~
/home/8cz/Workbench/MANTID/PD124_loggerAdjust_162_163/mantid/Framework/DataHandling/src/LoadEventNexus.cpp:552:33: note:                 ‘template<class T> static std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs(const string&, T, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<const Mantid::Kernel::TimeSeriesProperty<int> >&)’
  552 | std::shared_ptr<BankPulseTimes> LoadEventNexus::runLoadNexusLogs(
      |                                 ^~~~~~~~~~~~~~
```
So adding a default argument to a template function is not as clean and straight forward as one would assume. (see https://stackoverflow.com/questions/4050202/why-default-argument-cannot-be-specified-for-an-explicit-template-specialization for more info).
Also, the `unused argument` warning will come back and haunt us again on the build server as Mantid build server is configured to not allow a single unused args unless they are declared with `UNUSEDARG(arg)` in each function, which kind of defeat the whole efforts in avoiding side effects.
I hope this help explain why I choose to use overloading instead of default arguments for this particular function.

Reply:
The `-Wunused-parameter` doesn't seem very discriminatory :disappointed:.  
Well, if we can't have just one method, can you factor out the duplicate code  found in the implementation of methods 
```cpp
  template <typename T>
  static std::shared_ptr<BankPulseTimes> runLoadNexusLogs(
      const std::string &nexusfilename, T localWorkspace, Algorithm &alg,
      bool returnpulsetimes, int &nPeriods,
      std::unique_ptr<const Kernel::TimeSeriesProperty<int>> &periodLog);
```

and

```cpp
  template <typename T>
  static std::shared_ptr<BankPulseTimes> runLoadNexusLogs(
      const std::string &nexusfilename, T localWorkspace, Algorithm &alg,
      bool returnpulsetimes, int &nPeriods,
      std::unique_ptr<const Kernel::TimeSeriesProperty<int>> &periodLog,
      const std::vector<std::string> &allow_list,
      const std::vector<std::string> &block_list);
```
?

Reply:
Unfortunately I'm not proficient enough to make this happen.
Perhaps you could help me with it? 🙏

Reply:
I think I might a way, a very ugly way, but it might work

Reply:
The ugly solution seems to be working.  Could you give it another look?

Reply:
Unfortunately, this reduce code duplication endeavor is leading to a new doc-test error consistently.
```
Framework/WorkflowAlgorithms/test/CMakeFiles/WorkflowAlgorithmsTest.dir/__/__/TestHelpers/src/WorkspaceCreationHelper.cpp.o -o bin/WorkflowAlgorithmsTest  -Wl,-rpath,/jenkins_workdir/workspace/pull_requests-ubuntu/build/bin  /usr/lib/x86_64-linux-gnu/libboost_date_time.so  /usr/lib/x86_64-linux-gnu/libboost_regex.so  /usr/lib/x86_64-linux-gnu/libboost_serialization.so  /usr/lib/x86_64-linux-gnu/libboost_filesystem.so  /usr/lib/x86_64-linux-gnu/libboost_system.so  /usr/lib/libPocoFoundation.so  /usr/lib/libPocoUtil.so  /usr/lib/libPocoXML.so  /usr/lib/libPocoNet.so  /usr/lib/libPocoCrypto.so  /usr/lib/libPocoNetSSL.so  /usr/lib/x86_64-linux-gnu/libmuparser.so  bin/libMantidWorkflowAlgorithms.so  bin/libMantidAlgorithms.so  bin/libMantidDataHandling.so  bin/libMantidNexus.so  bin/libMantidDataObjects.so  bin/libMantidNexusGeometry.so  bin/libMantidAPI.so  bin/libMantidGeometry.so  bin/libMantidHistogramData.so  bin/libMantidIndexing.so  bin/libMantidParallel.so  bin/libMantidBeamline.so  bin/libMantidKernel.so  bin/libMantidTypes.so  /usr/lib/gcc/x86_64-linux-gnu/7/libgomp.so  -lpthread  /usr/lib/x86_64-linux-gnu/libboost_date_time.so  /usr/lib/x86_64-linux-gnu/libboost_regex.so  /usr/lib/x86_64-linux-gnu/libboost_serialization.so  /usr/lib/x86_64-linux-gnu/libboost_filesystem.so  /usr/lib/x86_64-linux-gnu/libboost_system.so  /usr/lib/libNeXus.so  /usr/lib/libNeXusCPP.so  /usr/lib/x86_64-linux-gnu/libjsoncpp.so  /usr/lib/x86_64-linux-gnu/libtbb.so  /usr/lib/x86_64-linux-gnu/libtbbmalloc.so  -Wl,-rpath-link,/jenkins_workdir/workspace/pull_requests-ubuntu/build/bin && :
12:48:26 bin/libMantidDataHandling.so: undefined reference to `std::shared_ptr<BankPulseTimes> Mantid::DataHandling::LoadEventNexus::runLoadNexusLogs<std::shared_ptr<Mantid::API::MatrixWorkspace> >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<Mantid::API::MatrixWorkspace>, Mantid::API::Algorithm&, bool, int&, std::unique_ptr<Mantid::Kernel::TimeSeriesProperty<int> const, std::default_delete<Mantid::Kernel::TimeSeriesProperty<int> const> >&)'
12:48:26 collect2: error: ld returned 1 exit status
```

Reply:
I gave up.  Somehow this fix is making things worse.  Revert back to original state with duplicated code in both functions.

Reply:
You gave it an honest try and that's that. I'll mark this as resolved